crates/artifact_content/src/raw_response_type.rs

```diff
@@ -0,0 +1,130 @@
+                        scalar_field.parent_object_entity_name
+                    ));
+                    continue;
+                }
+
+                let server_scalar_selectable = server_scalar_selectable_named(
+                    db,
+                    scalar_field.parent_object_entity_name,
+                    scalar_field.name.into(),
+                )
+                .as_ref()
+                .expect(
+                    "Expected validation to have succeeded. \
+                            This is indicative of a bug in Isograph.",
+                )
+                .as_ref()
+                .expect(
+                    "Expected selectable to exist. \
+                            This is indicative of a bug in Isograph.",
+                );
+
+                let raw_type = server_scalar_selectable.target_scalar_entity.as_ref().map(
+                    &mut |scalar_entity_name| match server_scalar_selectable
+                        .javascript_type_override
+                    {
+                        Some(javascript_name) => javascript_name,
+                        None => server_scalar_entity_javascript_name(db, *scalar_entity_name)
+                            .as_ref()
+                            .expect(
+                                "Expected parsing to not have failed. \
+                                        This is indicative of a bug in Isograph.",
+                            )
+                            .expect(
+                                "Expected entity to exist. \
+                                        This is indicative of a bug in Isograph.",
+                            ),
+                    },
+                );
+
+                let is_optional = matches!(
+                    server_scalar_selectable.target_scalar_entity,
+                    TypeAnnotation::Union(_)
+                );
+
+                raw_response_type.push_str(&format!(
+                    "{name}{}: {},\n",
+                    if is_optional { "?" } else { "" },
+                    print_javascript_type_declaration(&raw_type)
+                ));
+            }
+            MergedServerSelection::LinkedField(linked_field) => {
+                raw_response_type.push_str(indent);
+                let normalization_alias = linked_field.normalization_alias();
+                let name = normalization_alias
+                    .as_deref()
+                    .unwrap_or(linked_field.name.lookup());
+                raw_response_type.push_str(&format!("{name}: {{\n"));
+
+                generate_raw_response_type_inner(
+                    db,
+                    raw_response_type,
+                    &linked_field.selection_map,
+                    indentation_level + 1,
+                );
+                raw_response_type.push_str(&format!("{indent}}},\n"));
+            }
+            MergedServerSelection::ClientPointer(_) => {}
+            MergedServerSelection::InlineFragment(_) => {
```

Comment:
This would be true for GraphQL in general, but we don't support anything besides abstract -> concrete type refinements via `asConcreteType` fields, so we can generate a discriminated union here.